src/mlpack/methods/ann/rbm/binary_rbm_policy_impl.hpp

```diff
@@ -0,0 +1,167 @@
+#define MLPACK_METHODS_ANN_RBM_BINARY_RBM_POLICY_IMPL_HPP
+
+#include <mlpack/core.hpp>
+
+#include "binary_rbm_policy.hpp"
+
+namespace mlpack {
+namespace ann {
+
+template<typename DataType>
+BinaryRBMPolicy<DataType>
+    ::BinaryRBMPolicy(const size_t visibleSize, const size_t hiddenSize) :
+    visibleSize(visibleSize),
+    hiddenSize(hiddenSize)
+{
+  parameter.set_size((visibleSize * hiddenSize) + visibleSize + hiddenSize, 1);
+}
+
+template<typename DataType>
+void BinaryRBMPolicy<DataType>::Reset()
+{
+  weight = DataType(parameter.memptr(), hiddenSize, visibleSize, false, false);
+  hiddenBias = DataType(parameter.memptr() + weight.n_elem,
+      hiddenSize, 1, false, false);
+  visibleBias = DataType(parameter.memptr() + weight.n_elem +
+      hiddenBias.n_elem , visibleSize, 1, false, false);
+}
+
+template<typename DataType>
+typename BinaryRBMPolicy<DataType>::ElemType BinaryRBMPolicy<DataType>
+    ::FreeEnergy(DataType&& input)
+{
+  HiddenPreActivation(std::move(input), std::move(preActivation));
+  preActivation = arma::log(1 + arma::trunc_exp(preActivation));
+  return  -(arma::accu(preActivation) + arma::dot(input, visibleBias));
+}
+
+template<typename DataType>
+typename BinaryRBMPolicy<DataType>::ElemType BinaryRBMPolicy<DataType>
+    ::Evaluate(DataType& predictors, size_t i, size_t batchSize)
+{
+  size_t idx = RandInt(0, predictors.n_rows);
+  DataType temp = arma::round(predictors.col(i));
+  corruptInput = temp;
+  corruptInput.row(idx) = 1 - corruptInput.row(idx);
+  return std::log(LogisticFunction::Fn(FreeEnergy(std::move(corruptInput)) -
+      FreeEnergy(std::move(temp)))) * predictors.n_rows;
+}
+
+template<typename DataType>
+void BinaryRBMPolicy<DataType>::PositivePhase(
+    DataType&& input,
+    DataType&& gradient)
+{
+  DataType weightGrad = DataType(gradient.memptr(),
+      hiddenSize, visibleSize, false, false);
+
+  DataType hiddenBiasGrad = DataType(gradient.memptr() + weightGrad.n_elem,
+      hiddenSize, 1, false, false);
+
+  DataType visibleBiasGrad = DataType(gradient.memptr() + weightGrad.n_elem +
+      hiddenBiasGrad.n_elem, visibleSize, 1, false, false);
+
+  HiddenMean(std::move(input), std::move(hiddenBiasGrad));
+  weightGrad = hiddenBiasGrad * input.t();
+  visibleBiasGrad = input;
+}
+
+template<typename DataType>
+void BinaryRBMPolicy<DataType>::NegativePhase(
+    DataType&& negativeSamples,
+    DataType&& gradient)
+{
+  DataType weightGrad = DataType(gradient.memptr(),
+      hiddenSize, visibleSize, false, false);
+
+  DataType hiddenBiasGrad = DataType(gradient.memptr() + weightGrad.n_elem,
+      hiddenSize, 1, false, false);
+
+  DataType visibleBiasGrad = DataType(gradient.memptr() + weightGrad.n_elem +
+      hiddenBiasGrad.n_elem, visibleSize, 1, false, false);
+
+  HiddenMean(std::move(negativeSamples), std::move(hiddenBiasGrad));
+  weightGrad = hiddenBiasGrad * negativeSamples.t();
+  visibleBiasGrad = negativeSamples;
+}
```

Comment:
@lozhnikov `PositivePhase()` and `NegativePhase()` have the same implementation. Is there something wrong here?

Reply:
@Wenhao-H I think it's correct. I just didn't notice that:) In this case you can get rid of either `PositivePhase()` or `NegativePhase()`.